packages/react-router/lib/dom/ssr/single-fetch.tsx

```diff
@@ -359,36 +494,38 @@ async function singleFetchLoaderNavigationStrategy(
             return;
           }
 
-          // Otherwise, we opt out if we currently have data and a
+          // Otherwise, we opt out if we currently have data, a `loader`, and a
           // `shouldRevalidate` function.  This implies that the user opted out
           // via `shouldRevalidate`
           if (
             m.route.id in router.state.loaderData &&
-            manifestRoute &&
-            m.route.shouldRevalidate
+            hasLoader(m.route.id) &&
+            getShouldRevalidate(m.route.id)
+            // manifestRoute &&
+            // manifestRoute.hasLoader &&
+            // routeModules[m.route.id]?.shouldRevalidate
           ) {
-            if (manifestRoute.hasLoader) {
-              // If we have a server loader, make sure we don't include it in the
-              // single fetch .data request
-              foundOptOutRoute = true;
-            }
+            foundOptOutRoute = true;
```

Comment:
This is causing the failing E2E test - we need to short circuit either way, and then want to mark `foundOptOutRoute` only when `hasLoader()=true`

```ts
if (m.route.id in router.state.loaderData && getShouldRevalidate(m.route.id)) {
  if (hasLoader(m.route.id)) {
    // If we have a server loader, make sure we don't include it in the
    // single fetch .data request
    foundOptOutRoute = true;
  }
  return
}
```